src/common.rs

```diff
@@ -20,6 +23,75 @@ pub(crate) const fn const_byte_copy<
     }
 }
 
+#[cfg_attr(feature = "const_convert_and_const_trait_impl", const_trait)]
+pub trait Number: Sized + Copy + Clone + PartialOrd + Ord + PartialEq + Eq {
+    type UnderlyingType: Number
+        + Debug
+        + TryFrom<u8>
+        + TryFrom<u16>
+        + TryFrom<u32>
+        + TryFrom<u64>
+        + TryFrom<u128>;
+
+    /// Number of bits that can fit in this type
+    const BITS: usize;
+
+    /// Minimum value that can be represented by this type
+    const MIN: Self;
+
+    /// Maximum value that can be represented by this type
+    const MAX: Self;
+
+    /// Creates a number from the given value, throwing an error if the value is too large.
+    /// This constructor is useful when creating a value from a literal.
+    fn new(value: Self::UnderlyingType) -> Self;
+
+    /// Creates a number from the given value, return None if the value is too large
+    fn try_new(value: Self::UnderlyingType) -> Result<Self, TryNewError>;
+
+    fn value(self) -> Self::UnderlyingType;
+
+    /// Creates a number from the given value, throwing an error if the value is too large.
+    /// This constructor is useful when the value is convertible to T. Use [`Self::new`] for literals.
+    #[cfg(not(feature = "const_convert_and_const_trait_impl"))]
+    fn from_<T: Number>(value: T) -> Self;
+
+    /// Creates an instance from the given `value`. Unlike the various `new...` functions, this
+    /// will never fail as the value is masked to the result size.
+    #[cfg(not(feature = "const_convert_and_const_trait_impl"))]
+    fn masked_new<T: Number>(value: T) -> Self;
+
+    fn as_u8(&self) -> u8;
```

Comment:
Is there a reason these methods take a reference to `self` instead of an owned value? I'd expect the latter to be slightly more efficient for most types.

Reply:
I had assumed this would be optimized out anyway, given how small the type is anyway

Reply:
Sounds good, done (except for Number which is the compatibility trait)